File: java/app/src/main/java/com/ffo/mvp_java/mvp/base/BaseActivity.java
package com.ffo.mvp_java.mvp.base;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;

import com.ffo.mvp_java.mvp.IView;

/**
 * @author: huchunhua
 * @create_time: 2018/5/2
 * @change_time:
 * @package: com.ffo.mvp_java.mvp.base
 * @project: java
 * @mail: huchunhua5@wasu.com
 * @describe: 一句话描述
 */
public abstract class BaseActivity<P extends BasePresenter> extends Activity implements IView{
    protected P iPresenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        bindView();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        UnbindView();
    }

    @Override
    public void bindView() {
        iPresenter = createPresenter();
        iPresenter.register(this);
    }

    @Override
    public void UnbindView() {
        iPresenter.unRegister();
    }

    protected abstract P createPresenter();
}

File: kotlin/app/src/main/java/pffo/com/mvp/BaseActivity.kt
package com.ffo.ipet.mvp

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import com.ffo.ipet.mvp.contract.DemoContract
import com.ffo.ipet.mvp.imp.BasePresenter
import com.ffo.ipet.ui.activity.BaseActivity
import com.ffo.util.log.LogUtil

/**
 * @author: huchunhua
 * @create_time: 2018/4/26
 * @change_time:
 * @package: com.ffo.ipet.mvp
 * @project: pet
 * @mail: huchunhua5wasu.com
 * @describe: 一句话描述
 */
public open abstract class BaseActivity<V: IView , P : BasePresenter<V , V>>() : Activity(), IView {

    companion object {
        val log: LogUtil = LogUtil(BaseActivity::class.simpleName.toString())
    }

    protected lateinit var iPersenter: P

    protected var context: Context? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        log.infoMark("BaseActivity onCreate")
        bindView()
        context = this;
    }

    override fun onDestroy() {
        log.infoMark("BaseActivity onDestroy")
        context = null;

        super.onDestroy()
    }

    override fun getViewContext(): Context {
        return context ?: let { throw ExceptionInInitializerError("context is null") }
    }

    override fun bindView() {
        log.info("BaseActivity bindView")
        iPersenter = createPresenter()
        iPersenter.register(this)
    }

    override fun unBindView() {
        log.info("BaseActivity unBindView")
        iPersenter.unRegister()
    }

    /**
     * must init presenter
     */
    abstract fun createPresenter(): P;

}